### PR TITLE
quick fix for 2539

### PR DESCRIPTION
**Pull request by megaSpoon**
## What changes were proposed in this pull request?

fix issue 2539(https://github.com/intel-analytics/BigDL/issues/2539)
quick workaround. But the root cause is not revealed.

## Related links or issues (optional)
fix issue 2539(https://github.com/intel-analytics/BigDL/issues/2539)

### CODE REVIEW

File: pyspark/bigdl/util/engine.py [hunk 1]
```diff
@@ -67,8 +67,9 @@ def __prepare_bigdl_env():
 
     def append_path(env_var_name, path):
         try:
-            print("Adding %s to %s" % (path, env_var_name))
-            os.environ[env_var_name] = path + ":" + os.environ[env_var_name]  # noqa
+            if path not in os.environ[env_var_name]:
```
Comment (yiheng) [inline pyspark/bigdl/util/engine.py]: what if path is a substring of another in the env_var_name, e.g. your path is "/ABC/DEF", and the env_var_name is "/ABC/DEF/GHI:/DDD/XXX". The path is not in the env but your check will consider it in the env, right?

**Comment by yiheng** [comment]
Please add unit tests for your change.

**Comment by yiheng** [comment]
@hkiang01 any comments?

**Comment by hkiang01** [comment]
@yiheng I think you got the wrong guy

**Comment by yiheng** [comment]
@hkiang01 Sorry...

**Comment by yiheng** [comment]
@hkvision any comments?

**Comment by hkvision** [comment]
LGTM

**Comment by hkvision** [comment]
Just to be safe, can we check this as well when inserting into sys.path? i.e. https://github.com/megaSpoon/BigDL/blob/19437a436dc5772a71aa360fd9a7e5b94ab05998/pyspark/bigdl/util/engine.py#L59
https://github.com/megaSpoon/BigDL/blob/19437a436dc5772a71aa360fd9a7e5b94ab05998/pyspark/bigdl/util/engine.py#L82
https://github.com/megaSpoon/BigDL/blob/19437a436dc5772a71aa360fd9a7e5b94ab05998/pyspark/bigdl/util/engine.py#L90

File: pyspark/bigdl/util/engine.py (hunk 1 above)
Comment (hkvision) [reply pyspark/bigdl/util/engine.py]: What you doubt is reasonable. But in actual case, it is used to adding jar to SPARK_CLASSPATH, will this scenario happen?

Comment (yiheng) [reply pyspark/bigdl/util/engine.py]: that should be fine. But the code could not tell that. Should we change the path to jar_path to make it more clearer?

Comment (hkvision) [reply pyspark/bigdl/util/engine.py]: I'm fine for changing the name to `jar_path` as `append_path` is used for adding jars only.

Comment (megaSpoon) [reply pyspark/bigdl/util/engine.py]: we can solve it by checking path split by ":" to avoid the substring flaw.

**Comment by megaSpoon** [comment]
add codes to check insertion into sys.path and unit tests for it.

about the unit test:
The prepare_env() will be called when we import bigdl so I just called those functions in the engine which  append or insert path to the environment variables to see if they raise the expected exception.

File: pyspark/test/bigdl/test_engine_env.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import glob
+import pyspark
+
+
+class TestEngineEnv():
+    def setup_method(self, method):
+        """ setup any state tied to the execution of the given method in a
+        class.  setup_method is invoked for every test method of a class.
+        """
+        self.spark_home = '/home/megaspoon/Documents/spark-2.0.1-bin-hadoop2.7'
```
Comment (hkvision) [inline pyspark/test/bigdl/test_engine_env.py]: change the spark home?

File: pyspark/bigdl/util/engine.py [hunk 3]
```diff
@@ -41,6 +41,16 @@ def check_spark_source_conflict(spark_home, pyspark_path):
         warnings.warn(warning_msg)
 
 
+def __sys_path_insert(file_path):
+    try:
+        if file_path not in sys.path.split(":"):
+            print("Prepending %s to sys.path" % file_path)
+            sys.path.insert(0, file_path)
+    except:
```
Comment (hkvision) [inline pyspark/bigdl/util/engine.py]: Why will this raise an exception? In the try block even if file_path is in sys.path, I suppose it won't throw an error but just not go into the if branch?

Comment (hkvision) [reply pyspark/bigdl/util/engine.py]: I mean just add the if condition as well when inserting to sys.path...

File: pyspark/bigdl/util/engine.py [hunk 4]
```diff
@@ -41,6 +41,16 @@ def check_spark_source_conflict(spark_home, pyspark_path):
         warnings.warn(warning_msg)
 
 
+def __sys_path_insert(file_path):
+    try:
+        if file_path not in sys.path.split(":"):
```
[the remark this answers is not in the document]
Comment (megaSpoon) [reply pyspark/bigdl/util/engine.py]: Yep.

File: pyspark/test/bigdl/test_engine_env.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    def teardown_method(self, method):
+        """ teardown any state that was previously setup with a setup_method
+        call.
+        """
+        pass
+
+    def test___sys_path_insert(self):
+        py4j = glob.glob(os.path.join(self.spark_home, 'python/lib', 'py4j-*.zip'))[0]
+        pyspark = glob.glob(os.path.join(self.spark_home, 'python/lib', 'pyspark*.zip'))[0]
+        with pytest.raises(RuntimeError):
```
Comment (hkvision) [inline pyspark/test/bigdl/test_engine_env.py]: will this still raise an error?

File: pyspark/test/bigdl/test_engine_env.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        pass
+
+    def teardown_method(self, method):
+        """ teardown any state that was previously setup with a setup_method
+        call.
+        """
+        pass
+
+    def test___sys_path_insert(self):
+        py4j = glob.glob(os.path.join(self.spark_home, 'python/lib', 'py4j-*.zip'))[0]
```
Comment (hkvision) [inline pyspark/test/bigdl/test_engine_env.py]: is self.spark_home defined?

**Comment by yiheng** [comment]
@megaSpoon  any update of Kais's comments?

**Comment by megaSpoon** [comment]
Jenkins PR test #138 passed

**Pull request merged**